src/v/kafka/requests/fetch_request.cc

```diff
@@ -523,37 +530,60 @@ static ss::future<> fetch_topic_partition(
  * response must be reassembled such that the responses appear in these
  * order as the partitions in the request.
  */
-static ss::future<> fetch_topic_partitions(op_context& octx) {
-    return ss::do_with(
-      octx.response_begin(), [&octx](op_context::response_iterator& resp_it) {
-          return ss::do_for_each(
-                   octx.request.cbegin(),
-                   octx.request.cend(),
-                   [&octx, &resp_it](
-                     const fetch_request::const_iterator::value_type& p) {
-                       return fetch_topic_partition(octx, p, resp_it)
-                         .then([&resp_it] { resp_it++; });
-                   })
-            .then([&octx] { octx.reset_context(); });
+static ss::future<> fetch_topic_partitions(ss::lw_shared_ptr<op_context> octx) {
+    auto resp_it = octx->response_begin();
+    std::vector<ss::future<>> fetches;
+    std::transform(
+      octx->request.cbegin(),
+      octx->request.cend(),
+      std::back_inserter(fetches),
+      [&resp_it, octx](const fetch_request::const_iterator::value_type& p) {
+          octx->fetches_in_progress++;
+          // we use gate as we may not wait for all the fetch results
+          return ss::with_gate(octx->rctx.gate(), [octx, &p, &resp_it] {
```

Comment:
i'm not seeing how this works with not waiting on some of the fetches. `fetches` is a vector of futures and before returning from this method, either all of them are waited on or not. but in the later case, would we not cause seastar to throw exceptions about destroyed unresolved/unread futures?

Reply:
The futures will eventually be waited for by the `ss::with_gate`. We will explicitly wait for all the futures by calling `gate.close()`

Reply:
but isn't `std::vector<ss::future<>> fetches;` a stack variable? if execution takes the path that flows through `if (octx->fetches_in_progress > 0 && octx->request.debounce_delay()) {` then it appears that we may return from this method with non consumed futures in that vector. is that not possible, or is it the same as the `(void)xxx` pattern?

Reply:
this is exactly the `(void)...` pattern.

Reply:
got it. ah right its only ignore exceptional future that is the problem.